text/0000-repr-transparent.md

```diff
@@ -0,0 +1,191 @@
+
+```rust
+#[repr(transparent)]
+struct TransparentNewtype(f64);
+```
+
+This structure will still have the same representation as a raw `f64` value.
+
+Syntactically, the `repr` meta list will be extended to accept a meta item
+with the name "transparent". This attribute can be placed on newtypes,
+i.e. structures (and structure tuples) with a single field, and on structures
+that are logically equivalent to a newtype, i.e. structures with multiple fields
+where only a single one of them has a non-zero size.
+
+Some examples of `#[repr(transparent)]` are:
+
+```rust
+// Transparent struct tuple.
+#[repr(transparent)]
+struct TransparentStructTuple(i32);
+
+// Transparent structure.
+#[repr(transparent)]
+struct TransparentStructure { only_field: f64 }
+
+// Transparent struct wrapper with a marker.
+#[repr(transparent)]
+struct TransparentWrapper<T> {
+    only_non_zero_sized_field: f64,
+    marker: PhantomData<T>,
+}
+```
+
+This new representation is mostly useful when the structure it is put on must be
+used in FFI context as a wrapper to the underlying type without actually being
+affected by any ABI semantics.
+
+It is also useful for `AtomicUsize`-like types, which [RFC 1649] states should
+have the same representation as their underlying types.
+
+[RFC 1649]: https://github.com/rust-lang/rfcs/pull/1649
+
+This new representation cannot be used with any other representation attribute
+but alignment, to be able to specify a transparent wrapper with additional
+alignment constraints:
+
+```rust
+#[repr(transparent, align = "128")]
+struct OverAligned(f64); // Behaves as a bare f64 with 128 bits alignment.
+
+#[repr(C, transparent)]
+struct BogusRepr(f64); // Nonsensical, repr cannot be C and transparent.
+```
+
+As a matter of optimisation, eligible `#[repr(Rust)]` structs behave as if
+they were `#[repr(transparent)]` but as an implementation detail that can't be
+relied upon by users.
+
+```rust
+struct ImplicitlyTransparentWrapper(f64);
+
+#[repr(C)]
+struct BogusRepr {
+    // While ImplicitlyTransparentWrapper implicitly has the same representation
+    // as f64, this will fail to compile because ImplicitlyTransparentWrapper
+    // has no explicit transparent or C representation.
+    wrapper: ImplicitlyTransparentWrapper,
+}
+```
+
+The representation of a transparent wrapper is the representation of its
+only non-zero-sized field, transitively:
+
+```rust
+#[repr(transparent)]
+struct Transparent<T>(T);
+
+#[repr(transparent)]
+struct F64(f64);
+
+#[repr(C)]
+struct C(usize);
+
+type TransparentF64 = Transparent<F64>; // Behaves as f64.
+
+type TransparentString = Transparent<String>; // Representation is Rust.
+
+type TransparentC = Transparent<C>; // Representation is C.
+
+type TransparentTransparentC = Transparent<Transparent<C>>; // Transitively C.
+```
+
+Coercions and casting between the transparent wrapper and its non-zero-sized
+types are forbidden.
```

Comment:
What about `transmute`? Is it explicitly always legal to transmute between a `transparent` structure and its representation type, or between two `transparent` structures that have compatible representation types?

This is relevant because there are cases where today people are relying on the `repr(Rust)` "optimization" that allows such transmutes to be safe, but that optimization, as noted in this RFC, is not guaranteed. The presence of this attribute might be interpreted as changing that.

However, it also might be something that can be just as well left for a later RFC.